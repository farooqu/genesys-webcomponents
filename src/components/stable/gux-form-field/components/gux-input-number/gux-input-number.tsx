import { Component, Element, h, JSX, Prop, State } from '@stencil/core';

import { buildI18nForComponent, GetI18nValue } from '../../../../../i18n';
import setInputValue from '../../../../../utils/dom/set-input-value';
import simulateNativeEvent from '../../../../../utils/dom/simulate-native-event';
import { onDisabledChange } from '../../../../../utils/dom/on-attribute-change';

import componentResources from './i18n/en.json';

/**
 * @slot input - Required slot for input[type="radio"]
 */
@Component({
  styleUrl: 'gux-input-number.less',
  tag: 'gux-input-number'
})
export class GuxInputNumber {
  private input: HTMLInputElement;
  private getI18nValue: GetI18nValue;
  private disabledObserver: MutationObserver;

  @Element()
  private root: HTMLElement;

  @Prop()
  clearable: boolean;

  @State()
  private hasContent: boolean = false;

  @State()
  private disabled: boolean;

  private clearInput(): void {
    setInputValue(this.input, '', true);
  }

  private setHasContent(): void {
    this.hasContent = Boolean(this.input.value);
  }

  private simulateNativeInputAndChangeEvents(): void {
    simulateNativeEvent(this.input, 'input');
    simulateNativeEvent(this.input, 'change');
  }

  private stepUp(): void {
    if (this.input.value === '') {
      setInputValue(this.input, this.input.min || '0', false);
    } else {
      this.input.stepUp();
      this.simulateNativeInputAndChangeEvents();
    }
  }

  private stepDown(): void {
    if (this.input.value === '') {
      setInputValue(this.input, this.input.min || '0', false);
    } else {
      this.input.stepDown();
      this.simulateNativeInputAndChangeEvents();
    }
  }

  private renderClearButton(): JSX.Element {
    if (this.clearable && this.hasContent && !this.disabled) {
      return (
        <button
          class="gux-clear-button"
          tabIndex={-1}
          type="button"
          title={this.getI18nValue('clear')}
          disabled={this.disabled}
          onClick={this.clearInput.bind(this)}
        >
          <gux-icon icon-name="close" decorative></gux-icon>
        </button>
      ) as JSX.Element;
    }

    return null;
  }

  private renderStepButtons(): JSX.Element {
    return (
      <div class="gux-step-buttons-container">
        <button
          class="gux-step-button"
          tabIndex={-1}
          type="button"
          title={this.getI18nValue('increment')}
          disabled={this.disabled}
          onClick={() => this.stepUp()}
        >
          <gux-icon icon-name="chevron-small-up" decorative></gux-icon>
        </button>

        <button
          class="gux-step-button"
          tabIndex={-1}
          type="button"
          title={this.getI18nValue('decrement')}
          disabled={this.disabled}
          onClick={() => this.stepDown()}
        >
          <gux-icon icon-name="chevron-small-down" decorative></gux-icon>
        </button>
      </div>
    ) as JSX.Element;
  }

  async componentWillLoad(): Promise<void> {
    this.getI18nValue = await buildI18nForComponent(
      this.root,
      componentResources
    );

    this.input = this.root.querySelector('input[slot="input"]');

    this.setHasContent();
    this.disabled = this.input.disabled;

    this.input.addEventListener('input', () => {
      this.setHasContent();
    });
    this.disabledObserver = onDisabledChange(
      this.input,
      (disabled: boolean) => {
        this.disabled = disabled;
      }
    );
  }

  disconnectedCallback(): void {
    this.disabledObserver.disconnect();
  }

  render(): JSX.Element {
    return (
      <div
        class={{
          'gux-input-number-container': true,
          'gux-disabled': this.disabled
        }}
      >
        <div class="gux-input-container">
          <slot name="input" />
          {this.renderClearButton()}
        </div>
        {this.renderStepButtons()}
      </div>
    ) as JSX.Element;
  }
}
