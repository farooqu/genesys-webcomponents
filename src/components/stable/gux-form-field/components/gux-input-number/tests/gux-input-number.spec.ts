import { newSpecPage, SpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxInputNumber } from '../gux-input-number';

const components = [GuxInputNumber];
const language = 'en';

describe('gux-input-number', () => {
  let page: SpecPage;

  beforeEach(async () => {
    global.MutationObserver = MutationObserver;

    page = await newSpecPage({
      components,
      html: `
        <gux-input-number>
          <input type="number" slot="input"/>
        </gux-input-number>
      `,
      language
    });
  });

  it('should build', async () => {
    expect(page.rootInstance).toBeInstanceOf(GuxInputNumber);
  });

  it('should render', async () => {
    expect(page.root).toMatchSnapshot();
  });
});
