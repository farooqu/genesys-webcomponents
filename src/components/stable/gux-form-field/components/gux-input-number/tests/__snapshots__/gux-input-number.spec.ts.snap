// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-input-number should render 1`] = `
<gux-input-number>
  <!---->
  <div class="gux-input-number-container">
    <div class="gux-input-container">
      <input slot="input" type="number">
    </div>
    <div class="gux-step-buttons-container">
      <button class="gux-step-button" tabindex="-1" title="Increment" type="button">
        <gux-icon decorative="" icon-name="chevron-small-up"></gux-icon>
      </button>
      <button class="gux-step-button" tabindex="-1" title="Decrement" type="button">
        <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
      </button>
    </div>
  </div>
</gux-input-number>
`;
